### PR TITLE
Implemented Workshop Item Subscription

### PR DESCRIPTION
**Pull request by moldfield1995**
Added subscribing to items as well as the UGC interfaces
GetNumSubscribedItems and GetSubscribedItems. Edited SteamUGC Method
GetSubscribedItems to accept array instead of a ptr to allow c# calling

### CODE REVIEW

File: Facepunch.Steamworks/Interfaces/Workshop.cs [hunk 1]
```diff
@@ -123,6 +123,37 @@ public Item GetItem( ulong itemid )
...
+        {
+            return ugc.GetNumSubscribedItems();
+        }
+        /// <summary>
+        /// Creates a list of all item's that the user is currently subscribed
+        /// to for this App. This does not query titles or descriptions but
+        /// allows you to get the directories, state and ID of any subscribed Item.
+        /// This is mostly useful for getting all subscribed items install location.
+        /// </summary>
+        public Item[] GetSubscribedItems(uint amount)
```
Comment (garrynewman) [inline Facepunch.Steamworks/Interfaces/Workshop.cs]: I kind of doubt the usefulness of this function, plus why would you ever use the amount parameter?

File: Facepunch.Steamworks/SteamNative/SteamNative.SteamUGC.cs [hunk 2]
```diff
@@ -284,9 +284,12 @@ public bool GetQueryUGCStatistic( UGCQueryHandle_t handle /*UGCQueryHandle_t*/,
 		}
 		
 		// uint
-		public uint GetSubscribedItems( PublishedFileId_t* pvecPublishedFileID /*PublishedFileId_t **/, uint cMaxEntries /*uint32*/ )
+		public uint GetSubscribedItems( PublishedFileId_t[] pvecPublishedFileID /*PublishedFileId_t **/, uint cMaxEntries /*uint32*/ )
```
Comment (garrynewman) [inline Facepunch.Steamworks/SteamNative/SteamNative.SteamUGC.cs]: This file is all generated code, we shouldn't be editing it manually - but install fixing any problems with the generator (if any)

File: Facepunch.Steamworks/Interfaces/Workshop.cs [hunk 3]
```diff
@@ -123,6 +123,37 @@ public Item GetItem( ulong itemid )
             return new Item( itemid, this );
         }
 
+        /// <summary>
+        /// The amount of item's the user is subscribed to for this App
+        /// </summary>
+        public uint GetAmountSubsribed()
```
Comment (garrynewman) [inline Facepunch.Steamworks/Interfaces/Workshop.cs]: Spelling, should also probably be a property (although we could argue that there should just be a property of subscribed items, and you could just do Workshop.Subscriptions.Length)

**Comment by moldfield1995** [comment]
The GetSubscribedItems is rather useful in our project as its instantaneous and gets us all the directory's of each item so I can load them when the application is booting, where as other wise I would have to do a query (can some times take a bit)  or do some doing some oddness with going from the installed path to the workshop path. If you don't want this in the main version (because it is a bit of a edge case)I can just remove it for the main build. 
Thanks for the feedback

**Comment by garrynewman** [comment]
I feel like it's a redundant function, like there doesn't seem any reason we can't have the same instantly there behavior in a property that gets populated.

**Comment by skedgyedgy** [comment]
> I feel like it's a redundant function, like there doesn't seem any reason we can't have the same instantly there behavior in a property that gets populated.

-- i don't mean to revive a 6 year old thread but can we please reverse this decision??? it is not a redundant function as there is currently no way to do this in facepunch.steamworks (that i can find, at least) and this may actually prevent me from switching over. this is like core functionality for steam workshop, being able to access the client-sided list of subscribed addons instantly without having to do a whole query. not to mention that not having access to this makes info that would otherwise be accessible at any time now online-only

**Pull request merged**